Ethernet driver for native posix board

jukkar opened the pull request:
This creates ethernet TAP driver for native posix board. The driver creates and setups host TAP device (by default **zeth**) which can be used to communicate between zephyr and host.

jukkar committed:
drivers: eth: Add ethernet driver for native posix arch

This creates zeth network interface in your host and allows user
to send and receive data sent to this network interface.

Fixes #6007

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: eth_native_posix: Sample for native networking

This application just enables native host networking and net-shell
which allows connectivity from zephyr to host system via tap device.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

aescolar-ot commented on samples/net/eth_native_posix/net_setup_host:
```diff
@@ -0,0 +1,59 @@
+   ip -6 address add $IPV6_ADDR_1 dev $IFACE
+fi
+
+if [ ! -z "$IPV4_ROUTE_1" ]; then
```
IPV6?

jukkar replied:
Indeed, will fix this typo

aescolar-ot commented on drivers/ethernet/CMakeLists.txt:
```diff
@@ -6,3 +6,7 @@ zephyr_sources_ifdef(CONFIG_ETH_DW		eth_dw.c)
 zephyr_sources_ifdef(CONFIG_ETH_ENC28J60	eth_enc28j60.c)
 zephyr_sources_ifdef(CONFIG_ETH_MCUX		eth_mcux.c)
 zephyr_sources_ifdef(CONFIG_ETH_STM32_HAL	eth_stm32_hal.c)
+zephyr_sources_ifdef(CONFIG_ETH_NATIVE_POSIX
```
Would you consider adding a few lines for this driver in http://docs.zephyrproject.org/boards/posix/native_posix/doc/board.html#peripherals ?

jukkar replied:
Sure, np

aescolar-ot commented on samples/net/eth_native_posix/README.rst:
```diff
@@ -0,0 +1,35 @@
+
+.. code-block:: console
+
+    $ sudo --preserve-env=ZEPHYR_BASE make -s -C outdir/native_posix run
```
Maybe a silly question: Would it be possible to run 2 instances of this application back to back?

jukkar replied:
Absolutely, one could have multiple config files with properly prepared IP addresses in them. I will add command line options passing to default setup file (net_host_setup) that will help to meet this goal.

aescolar-ot commented on drivers/ethernet/Kconfig.native_posix:
```diff
@@ -0,0 +1,83 @@
+
+config ETH_NATIVE_POSIX_DEV_NAME
+	string "Host ethernet TUN/TAP device name"
+	default "/dev/net/tun"
```
Just a comment: It would seem to me that a few of these options would benefit from a command line switch to set them (or override their default). When #6138 is merged it should be easy to add them.

jukkar replied:
Sure, makes sense. I would like to keep also possibility to configure this via zephyr config file too.

aescolar-ot commented on drivers/ethernet/eth_native_posix.c:
```diff
@@ -0,0 +1,268 @@
+	ctx->mac_addr[2] = 0x5E;
+	ctx->mac_addr[3] = 0x00;
+	ctx->mac_addr[4] = 0x53;
+	ctx->mac_addr[5] = sys_rand32_get();
```
Just a silly note: Beware that until #6138 and #6128 are merged you will be getting exactly the same random number here all the time.

jukkar replied:
That is fine, and if needed, user can specify unique mac address in .config file.

aescolar-ot commented on samples/net/eth_native_posix/README.rst:
```diff
@@ -0,0 +1,35 @@
+   :compact:
+
+Normally one needs extra priviliges to create and configure the TAP device in
+host system. So you would need to execute the Zephyr process with admin
```
+the +host system[...]?

dbkinder replied:
``in the host system.  Use sudo to execute the Zephyr process with admin privileges, like this:``

dbkinder commented on samples/net/eth_native_posix/README.rst:
```diff
@@ -0,0 +1,35 @@
+Overview
+********
+
+The eth_native_posix sample application for Zephyr creates **zeth** network
```
``creates a **zeth**``

dbkinder commented on samples/net/eth_native_posix/README.rst:
```diff
@@ -0,0 +1,35 @@
+********
+
+The eth_native_posix sample application for Zephyr creates **zeth** network
+interface to host system. One can communicate with Zephyr via this network
```
``to the host``

aescolar-ot commented on drivers/ethernet/Kconfig.native_posix:
```diff
@@ -0,0 +1,83 @@
+	help
+	  This option sets the TUN/TAP device name in your host system.
+
+config ETH_NATIVE_POSIX_RX_STACK_SIZE
```
Is this a zephyr thread stack size? if so you should be able to just set it to ARCH_POSIX_RECOMMENDED_STACK_SIZE without the need for this config option.

jukkar replied:
Ok, I will remove this option.

aescolar-ot commented on drivers/ethernet/eth_native_posix_adapt.c:
```diff
@@ -0,0 +1,135 @@
+ * @file
+ *
+ * Routines setting up the host system. Those are placed in separate file
+ * because there is naming conflicts between host and zephyr network stacks.
```
Which ones are conflicting?

jukkar replied:
There are network related symbols (like struct sockaddr_in6) in include/net/net_ip.h that will conflict with host /usr/include.

aescolar reviewed (commented):
Thanks Jukka. Here a few comments after a relatively quick look thru.

lpereira commented on drivers/ethernet/eth_native_posix_adapt.c:
```diff
@@ -0,0 +1,135 @@
+	return -EAGAIN;
+}
+
+int eth_read_data(int fd, u8_t *buf, int buf_len)
```
Shouldn't this return a `ssize_t`?

tbursztyka replied:
indeed, same for write below

lpereira commented on drivers/ethernet/eth_native_posix.c:
```diff
@@ -0,0 +1,268 @@
+		CONFIG_ETH_NATIVE_POSIX_RX_STACK_SIZE);
+
+	while (1) {
+		if (net_if_is_up(ctx->iface)) {
```
Instead of polling all interfaces, why not select() on all of them at once?

tbursztyka replied:
the driver does not support yet multiple iface it seems (line 57)

jukkar replied:
Hmm, I am not sure I understood the question. Currently this eth driver supports only one TAP device. If you mean that we could have NULL timeout when calling select(), then that would not work as then other part of the system would not get chance to run.

jukkar replied:
> the driver does not support yet multiple iface it seems (line 57)

No, the VLAN PR at #5688 would need to go in first as it has various things to support multiple ethernet interfaces. This driver can then enhanced to support multiple interfaces (I actually had TODO comment in the beginning of the file)

tbursztyka commented on drivers/ethernet/eth_native_posix.c:
```diff
@@ -0,0 +1,268 @@
+}
+
+#if !defined(CONFIG_ETH_NATIVE_POSIX_RANDOM_MAC)
+static int mac_addr_from_str(struct eth_context *ctx, const char *src)
```
I think it will be wise to put that function somewhere in a unique place, slip uses it also.

jukkar replied:
Perhaps, I will investigate this more.

tbursztyka commented on drivers/ethernet/eth_native_posix.c:
```diff
@@ -0,0 +1,268 @@
+	int ret;
+
+	ret = eth_read_data(fd, ctx->recv, sizeof(ctx->recv));
+	if (ret > 0) {
```
instead, could you test for error (or 0) and return there?

It would indent less the below code

jukkar replied:
Sure, why not

jukkar commented on drivers/ethernet/Kconfig.native_posix:
```diff
@@ -0,0 +1,83 @@
+	  data from TAP device and passing that data to zephyr.
+
+config	SYS_LOG_ETH_NATIVE_POSIX_LEVEL
+	int
```
Oh yes, that is indeed better. I missed the generic option.

jukkar committed:
net: Create helper to convert MAC strings to array of bytes

As following commits need this functionality, create a function
which converts "01:02:ab:fe:34:dd" type hex strings to array of
bytes. Change the SLIP driver to use this new function.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Updated the PR according to comments. Also added instructions to README.rst file how to connect two zephyr instances together.

jukkar commented:
Fixed the compile error if debugging was enabled.

dbkinder reviewed (approved):
Thanks

jukkar commented:
New version uploaded. Tested now with Fedora 27.

aescolar-ot commented on samples/net/eth_native_posix/README.rst:
```diff
@@ -0,0 +1,112 @@
+
+.. code-block:: console
+
+    sudo --preserve-env=ZEPHYR_BASE make run
```
Seems this feature came in sudo 1.8.21 ( https://www.sudo.ws/stable.html#1.8.21 ) which is quite new (2017-08-23)

jukkar replied:
Ok, then we could just use -E option here. Fortunately this sudo cmd is only mentioned/used in readme file so no changes needed in the code.

dbkinder replied:
Are you going to remove the ``--preserve-env=ZEPHYR_BASE`` here (and just use ``--preserve-env``, and in the command examples below?

jukkar replied:
I was hoping to leave the --preserve-env=ZEPHYR_BASE version as that is the preferred way. But as the older sudo version does not support that I added also a way to run the sudo the old way. I did not touch the other sudo calls in that document as I felt that readers are "smart" enough to understand how to run the command.
I have no big preference with the sudo command parameters thou, so if you think that we should just use --preserve-env here, then I can just send a version using just that.

dbkinder replied:
Given the comments above, it seems we should just use ``--preserve-env``. But if you'd like to leave the ``preserve-env=ZEPHYR_BASE`` in, add the note mentioned below to the code example.

aescolar-ot replied:
Another topic on the same line:
You are calling here `make run` instead of just running the binary (`zephyr/zephyr.exe`)
Would it not be better to call the binary directly? (so we don't hide from the user what is happening) (I'm thinking that the user will probably want to run the binary in a debugger, pass options to it, or instrument it. So it would seem better to make it evident to them what is the binary they actually run)

jukkar commented:
Rebased, and added more information about sudo command possible failure to sample README file.

dbkinder commented on samples/net/eth_native_posix/README.rst:
```diff
@@ -0,0 +1,118 @@
+    mkdir -p build1/native_posix
+    cmake -DCONF_FILE=prj1.conf -DBOARD=native_posix -Bbuild1/native_posix -H.
+    make -s -C build1/native_posix
+    sudo --preserve-env=ZEPHYR_BASE make -s -C build1/native_posix run
```
just ``--preserve-env`` now?

dbkinder commented on samples/net/eth_native_posix/README.rst:
```diff
@@ -0,0 +1,118 @@
+    mkdir -p build2/native_posix
+    cmake -DCONF_FILE=prj2.conf -DBOARD=native_posix -Bbuild2/native_posix -H.
+    make -s -C build2/native_posix
+    sudo --preserve-env=ZEPHYR_BASE make -s -C build2/native_posix run
```
just ``--preserve-env`` now?

dbkinder commented on samples/net/eth_native_posix/README.rst:
```diff
@@ -0,0 +1,118 @@
+    CONFIG_ETH_NATIVE_POSIX_SETUP_SCRIPT="echo"
+    CONFIG_ETH_NATIVE_POSIX_DRV_NAME="zeth.2"
+
+Then compile and run two Zephyr instances:
```
If you leave the example as is, then add a note here:

    Then compile and run two Zephyr instances (if ``sudo --preserve-env=ZEPHYR_BASE`` 
    gives an error, just use ``sudo --preserve-env`` instead):

jukkar commented:
Readme file updated according to comments.

jukkar commented:
@dbkinder Are you ok with the PR documentation?

aescolar-ot commented on boards/posix/native_posix/doc/board.rst:
```diff
@@ -389,6 +389,14 @@ The following peripherals are currently provided with this board:
   unmasked. SW interrupts are also supported.
 
+**Ethernet driver**:
+  A simple TAP based ethernet driver is provided. The driver will create
```
One minor comment more from my side: Would it make sense to recommend users to ensure they set NATIVE_POSIX_SLOWDOWN_TO_REAL_TIME when using this driver?

nashif replied:
NATIVE_POSIX_SLOWDOWN_TO_REAL_TIME is enabled only for tests, so we should be good I guess.
